File: phases/19-capstone-projects/01-terminal-native-coding-agent/code/main.ts
// Capstone 19/01: terminal-native coding agent harness skeleton (TypeScript).
//
// Sources:
//   This lesson's docs/en.md (the Bun + Ink TUI harness with eight 2026 hooks)
//   Claude Code docs            https://docs.anthropic.com/en/docs/claude-code
//   Model Context Protocol      https://blog.modelcontextprotocol.io/posts/2026-mcp-roadmap/
//   OpenTelemetry GenAI semconv https://opentelemetry.io/docs/specs/semconv/gen-ai/
//
// The harness side of the capstone: REPL command parser, scripted (offline) LLM,
// tool dispatcher with read_file/run_shell, eight-event hook bus, plan state
// rewritten whole each turn, hard ceilings on turns/tokens/dollars, plus a tiny
// pass/fail eval counter. Zero network, stdlib only.
//
// Run: npx -y tsx@4.19.2 code/main.ts

import { readFileSync } from "node:fs";
import * as path from "node:path";
import * as readline from "node:readline";

type Status = "pending" | "in_progress" | "done" | "failed";

type TodoItem = {
  id: number;
  description: string;
  status: Status;
  note: string;
};

class PlanState {
  goal: string;
  items: TodoItem[];

  constructor(goal: string) {
    this.goal = goal;
    this.items = [];
  }

  rewrite(items: TodoItem[]): void {
    this.items = items;
  }

  summary(): string {
    const mark: Record<Status, string> = {
      pending: " ",
      in_progress: ">",
      done: "x",
      failed: "!",
    };
    const lines = [`GOAL: ${this.goal}`];
    for (const it of this.items) {
      lines.push(`  [${mark[it.status]}] ${it.id}. ${it.description}`);
    }
    return lines.join("\n");
  }
}

class Budget {
  maxTurns = 50;
  maxTokens = 200_000;
  maxDollars = 5.0;
  turnsUsed = 0;
  tokensUsed = 0;
  dollarsUsed = 0;

  step(tokens: number, dollars: number): void {
    this.turnsUsed += 1;
    this.tokensUsed += tokens;
    this.dollarsUsed += dollars;
  }

  exceeded(): string | null {
    if (this.turnsUsed >= this.maxTurns) return "turn_limit";
    if (this.tokensUsed >= this.maxTokens) return "token_limit";
    if (this.dollarsUsed >= this.maxDollars) return "dollar_limit";
    return null;
  }

  snapshot(): { turnsUsed: number; tokensUsed: number; dollarsUsed: number } {
    return {
      turnsUsed: this.turnsUsed,
      tokensUsed: this.tokensUsed,
      dollarsUsed: this.dollarsUsed,
    };
  }
}

type HookEvent =
  | "SessionStart"
  | "SessionEnd"
  | "PreToolUse"
  | "PostToolUse"
  | "UserPromptSubmit"
  | "Notification"
  | "Stop"
  | "PreCompact";

type HookPayload = Record<string, unknown>;
type HookFn = (payload: HookPayload) => HookPayload;

class HookBus {
  static readonly EVENTS: HookEvent[] = [
    "SessionStart",
    "SessionEnd",
    "PreToolUse",
    "PostToolUse",
    "UserPromptSubmit",
    "Notification",
    "Stop",
    "PreCompact",
  ];

  private hooks: Map<HookEvent, HookFn[]> = new Map();

  constructor() {
    for (const e of HookBus.EVENTS) this.hooks.set(e, []);
  }

  on(event: HookEvent, fn: HookFn): void {
    this.hooks.get(event)!.push(fn);
  }

  fire(event: HookEvent, payload: HookPayload): HookPayload {
    let current = payload;
    for (const fn of this.hooks.get(event)!) {
      current = fn(current) ?? current;
    }
    return current;
  }
}

const TRUNCATE_BYTES = 4096;

type ToolArgs = Record<string, string>;
type ToolFn = (sandbox: string, args: ToolArgs) => string;

function toolReadFile(sandbox: string, args: ToolArgs): string {
  const target = args.path ?? "";
  const full = path.resolve(sandbox, target);
  const root = path.resolve(sandbox);
  if (!full.startsWith(root + path.sep) && full !== root) {
    throw new Error("path escapes sandbox");
  }
  const data = readFileSync(full, "utf8");
  return data.slice(0, TRUNCATE_BYTES);
}

function toolRunShell(_sandbox: string, args: ToolArgs): string {
  // Offline scaffold: stub the shell. The capstone wires this to E2B / Daytona.
  // Returning a deterministic, truncated stand-in keeps the loop testable.
  const cmd = args.cmd ?? "";
  const stub: Record<string, string> = {
    ls: "README.md\nmain.ts\nmain.py",
    "git status": "On branch agent/demo\nnothing to commit, working tree clean",
  };
  const out = stub[cmd] ?? `(stub) ran: ${cmd}`;
  return `exit=0\n${out.slice(0, TRUNCATE_BYTES)}`;
}

const TOOLS: Record<string, ToolFn> = {
  read_file: toolReadFile,
  run_shell: toolRunShell,
};

type ToolCall = { name: string; args: ToolArgs };

type ModelTurn = {
  plan: TodoItem[];
  tool: ToolCall | null;
  tokens: number;
  cost: number;
};

type ScriptStep = {
  plan: ReadonlyArray<readonly [string, Status]>;
  tool: ToolCall | null;
  tokens: number;
  cost: number;
};

const SCRIPT: ScriptStep[] = [
  {
    plan: [
      ["locate target file", "in_progress"],
      ["read and diagnose", "pending"],
      ["apply fix and verify", "pending"],
    ],
    tool: { name: "run_shell", args: { cmd: "ls" } },
    tokens: 1200,
    cost: 0.02,
  },
  {
    plan: [
      ["locate target file", "done"],
      ["read and diagnose", "in_progress"],
      ["apply fix and verify", "pending"],
    ],
    tool: { name: "read_file", args: { path: "main.ts" } },
    tokens: 900,
    cost: 0.02,
  },
  {
    plan: [
      ["locate target file", "done"],
      ["read and diagnose", "done"],
      ["apply fix and verify", "done"],
    ],
    tool: null,
    tokens: 600,
    cost: 0.01,
  },
];

class ScriptedModel {
  step(_plan: PlanState, turn: number): ModelTurn {
    if (turn >= SCRIPT.length) {
      return { plan: [], tool: null, tokens: 200, cost: 0.005 };
    }
    const s = SCRIPT[turn];
    const items: TodoItem[] = s.plan.map(([description, status], i) => ({
      id: i + 1,
      description,
      status,
      note: "",
    }));
    return { plan: items, tool: s.tool, tokens: s.tokens, cost: s.cost };
  }
}

function destructiveGuard(payload: HookPayload): HookPayload {
  const args = (payload.args ?? {}) as ToolArgs;
  const cmd = args.cmd ?? "";
  if (cmd.includes("rm -rf") || cmd.includes("shutdown")) {
    return { ...payload, blocked: true, reason: "destructive command blocked by PreToolUse hook" };
  }
  return payload;
}

type RunResult = {
  plan: string;
  budget: { turnsUsed: number; tokensUsed: number; dollarsUsed: number };
  trace: HookPayload[];
  passed: boolean;
};

function runAgent(task: string, sandbox: string): RunResult {
  const plan = new PlanState(task);
  const budget = new Budget();
  const hooks = new HookBus();
  const trace: HookPayload[] = [];
  const model = new ScriptedModel();

  hooks.on("PreToolUse", destructiveGuard);
  hooks.on("PostToolUse", (p) => {
    trace.push({ event: "tool", ...p });
    return p;
  });
  hooks.on("SessionStart", (p) => {
    trace.push({ event: "start", ...p });
    return p;
  });
  hooks.on("SessionEnd", (p) => {
    trace.push({ event: "end", ...p });
    return p;
  });
  hooks.on("Stop", (p) => {
    trace.push({ event: "stop", ...p });
    return p;
  });

  hooks.fire("SessionStart", { task, sandbox, startedAt: Date.now() });

  let turn = 0;
  let completed = false;
  while (true) {
    const limit = budget.exceeded();
    if (limit) {
      hooks.fire("Stop", { reason: limit, turn });
      break;
    }
    const step = model.step(plan, turn);
    plan.rewrite(step.plan);
    budget.step(step.tokens, step.cost);

    if (step.tool === null) {
      hooks.fire("Stop", { reason: "complete", turn });
      completed = true;
      break;
    }

    const { name, args } = step.tool;
    const pre = hooks.fire("PreToolUse", { tool: name, args });
    if (pre.blocked) {
      hooks.fire("PostToolUse", {
        tool: name,
        blocked: true,
        reason: String(pre.reason ?? ""),
      });
      turn += 1;
      continue;
    }

    try {
      const result = TOOLS[name](sandbox, args);
      hooks.fire("PostToolUse", { tool: name, ok: true, bytes: result.length });
    } catch (err) {
      const e = err as Error;
      hooks.fire("PostToolUse", { tool: name, ok: false, error: e.message });
    }
    turn += 1;
  }

  hooks.fire("SessionEnd", budget.snapshot() as unknown as HookPayload);

  const allDone = plan.items.length > 0 && plan.items.every((it) => it.status === "done");
  return {
    plan: plan.summary(),
    budget: budget.snapshot(),
    trace,
    passed: completed && allDone,
  };
}

type Command =
  | { kind: "run"; task: string }
  | { kind: "eval" }
  | { kind: "help" }
  | { kind: "quit" }
  | { kind: "unknown"; raw: string };

function parseCommand(line: string): Command {
  const trimmed = line.trim();
  if (!trimmed) return { kind: "help" };
  if (trimmed === "quit" || trimmed === "exit") return { kind: "quit" };
  if (trimmed === "help" || trimmed === "?") return { kind: "help" };
  if (trimmed === "eval") return { kind: "eval" };
  const m = /^run\s+(.+)$/.exec(trimmed);
  if (m) return { kind: "run", task: m[1] };
  return { kind: "unknown", raw: trimmed };
}

function helpText(): string {
  return [
    "harness commands:",
    "  run <task>   plan/act/observe loop for one task against the scripted model",
    "  eval         run the offline eval and print pass/fail counts",
    "  help         show this message",
    "  quit         exit",
  ].join("\n");
}

const EVAL_TASKS: { task: string; expectedDone: number }[] = [
  { task: "diagnose worker.rs", expectedDone: 3 },
  { task: "summarize README", expectedDone: 3 },
  { task: "run smoke tests", expectedDone: 3 },
];

function runEval(sandbox: string): { passed: number; failed: number } {
  let passed = 0;
  let failed = 0;
  for (const t of EVAL_TASKS) {
    const r = runAgent(t.task, sandbox);
    const doneCount = (r.plan.match(/\[x\]/g) ?? []).length;
    if (r.passed && doneCount >= t.expectedDone) passed += 1;
    else failed += 1;
  }
  return { passed, failed };
}

function isInteractive(): boolean {
  return process.stdin.isTTY === true && process.argv.includes("--repl");
}

async function repl(sandbox: string): Promise<void> {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  console.log(helpText());
  const ask = (prompt: string): Promise<string> =>
    new Promise((resolve) => rl.question(prompt, resolve));
  while (true) {
    const line = await ask("agent> ");
    const cmd = parseCommand(line);
    if (cmd.kind === "quit") break;
    if (cmd.kind === "help") {
      console.log(helpText());
      continue;
    }
    if (cmd.kind === "eval") {
      const e = runEval(sandbox);
      console.log(`eval: passed=${e.passed} failed=${e.failed}`);
      continue;
    }
    if (cmd.kind === "run") {
      const r = runAgent(cmd.task, sandbox);
      console.log(r.plan);
      console.log("---");
      console.log(
        `turns=${r.budget.turnsUsed} tokens=${r.budget.tokensUsed} ` +
          `dollars=$${r.budget.dollarsUsed.toFixed(3)} passed=${r.passed}`,
      );
      continue;
    }
    console.log(`unknown command: ${cmd.raw}; type 'help'`);
  }
  rl.close();
}

async function main(): Promise<void> {
  const sandbox = path.resolve(__dirname);
  if (isInteractive()) {
    await repl(sandbox);
    return;
  }
  const task = "demonstrate the plan-act-observe loop without network calls";
  const result = runAgent(task, sandbox);
  console.log(result.plan);
  console.log("---");
  console.log(
    `turns=${result.budget.turnsUsed} tokens=${result.budget.tokensUsed} ` +
      `dollars=$${result.budget.dollarsUsed.toFixed(3)}`,
  );
  console.log("---");
  console.log(`trace events: ${result.trace.length}`);
  for (const ev of result.trace) console.log(" ", JSON.stringify(ev));
  console.log("---");
  const e = runEval(sandbox);
  console.log(`eval: passed=${e.passed} failed=${e.failed}`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: phases/19-capstone-projects/01-terminal-native-coding-agent/code/main.py
"""ターミナルネイティブ Coding Agent の最小 plan/act/observe loop scaffold。

2026年の coding agent で難しい architectural primitive は model call や
単体の tool ではありません。bounded context、structured plan state、
sandboxed tool dispatcher、各 lifecycle point の hook callback を備えた
plan-act-observe-recover loop です。このファイルはその loop を stdlib Python
だけで end to end に実装します。LLM は deterministic script で stub し、
network call なしでも loop logic を観測・test できるようにしています。

Run:  python main.py
"""

from __future__ import annotations

import json
import os
import subprocess
import sys
import time
from dataclasses import asdict, dataclass, field
from typing import Any, Callable


# ---------------------------------------------------------------------------
# plan state  --  TodoWrite 形状。毎 turn 全体を書き換える
# ---------------------------------------------------------------------------

@dataclass
class TodoItem:
    id: int
    description: str
    status: str  # "pending" | "in_progress" | "done" | "failed"
    note: str = ""


@dataclass
class PlanState:
    goal: str
    items: list[TodoItem] = field(default_factory=list)

    def summary(self) -> str:
        lines = [f"目標: {self.goal}"]
        for it in self.items:
            mark = {"pending": " ", "in_progress": ">", "done": "x", "failed": "!"}[it.status]
            lines.append(f"  [{mark}] {it.id}. {it.description}")
        return "\n".join(lines)


# ---------------------------------------------------------------------------
# budget  --  turns, tokens, dollars の hard ceiling
# ---------------------------------------------------------------------------

@dataclass
class Budget:
    max_turns: int = 50
    max_tokens: int = 200_000
    max_dollars: float = 5.00
    turns_used: int = 0
    tokens_used: int = 0
    dollars_used: float = 0.0

    def step(self, tokens: int, dollars: float) -> None:
        self.turns_used += 1
        self.tokens_used += tokens
        self.dollars_used += dollars

    def exceeded(self) -> str | None:
        if self.turns_used >= self.max_turns:
            return "turn_limit"
        if self.tokens_used >= self.max_tokens:
            return "token_limit"
        if self.dollars_used >= self.max_dollars:
            return "dollar_limit"
        return None


# ---------------------------------------------------------------------------
# hooks  --  2026年版の8 event surface (Pre/PostToolUse, SessionStart/End など)
# ---------------------------------------------------------------------------

HookFn = Callable[[dict[str, Any]], dict[str, Any]]


class HookBus:
    EVENTS = ("SessionStart", "SessionEnd", "PreToolUse", "PostToolUse",
              "UserPromptSubmit", "Notification", "Stop", "PreCompact")

    def __init__(self) -> None:
        self._hooks: dict[str, list[HookFn]] = {e: [] for e in self.EVENTS}

    def on(self, event: str, fn: HookFn) -> None:
        self._hooks[event].append(fn)

    def fire(self, event: str, payload: dict[str, Any]) -> dict[str, Any]:
        for fn in self._hooks[event]:
            payload = fn(payload) or payload
        return payload


# ---------------------------------------------------------------------------
# tool surface  --  各 tool は sandbox 化され、truncated text を返す
# ---------------------------------------------------------------------------

TRUNCATE_BYTES = 4096


def tool_read_file(sandbox: str, path: str) -> str:
    full = os.path.join(sandbox, path)
    if not os.path.realpath(full).startswith(os.path.realpath(sandbox)):
        raise RuntimeError("path が sandbox 外へ出ています")
    with open(full, "r", encoding="utf-8", errors="replace") as fh:
        return fh.read()[:TRUNCATE_BYTES]


def tool_run_shell(sandbox: str, cmd: str, timeout: int = 30) -> str:
    proc = subprocess.run(cmd, cwd=sandbox, shell=True, capture_output=True,
                          text=True, timeout=timeout)
    out = (proc.stdout + proc.stderr)[:TRUNCATE_BYTES]
    return f"exit={proc.returncode}\n{out}"


TOOLS: dict[str, Callable[..., str]] = {
    "read_file": tool_read_file,
    "run_shell": tool_run_shell,
}


# ---------------------------------------------------------------------------
# stub model  --  LLM なしで loop を test できる deterministic script
# ---------------------------------------------------------------------------

SCRIPT = [
    {"plan": [("対象ファイルを探す", "in_progress"),
              ("読み取って診断する", "pending"),
              ("修正を適用して検証する", "pending")],
     "tool": ("run_shell", {"cmd": "ls"}),
     "tokens": 1200, "cost": 0.02},
    {"plan": [("対象ファイルを探す", "done"),
              ("読み取って診断する", "in_progress"),
              ("修正を適用して検証する", "pending")],
     "tool": ("read_file", {"path": "README.md"}),
     "tokens": 900, "cost": 0.02},
    {"plan": [("対象ファイルを探す", "done"),
              ("読み取って診断する", "done"),
              ("修正を適用して検証する", "done")],
     "tool": None,  # 終了 turn
     "tokens": 600, "cost": 0.01},
]


def model_step(plan: PlanState, turn: int) -> dict[str, Any]:
    """stub model: plan rewrite と必要なら tool call を返す。"""
    if turn >= len(SCRIPT):
        return {"plan": plan.items, "tool": None, "tokens": 200, "cost": 0.005}
    s = SCRIPT[turn]
    items = [TodoItem(i + 1, desc, status) for i, (desc, status) in enumerate(s["plan"])]
    return {"plan": items, "tool": s["tool"], "tokens": s["tokens"], "cost": s["cost"]}


# ---------------------------------------------------------------------------
# main loop  --  hook 統合付き plan / act / observe / recover
# ---------------------------------------------------------------------------

def destructive_guard(payload: dict[str, Any]) -> dict[str, Any]:
    cmd = payload.get("args", {}).get("cmd", "")
    if "rm -rf" in cmd or "shutdown" in cmd:
        payload["blocked"] = True
        payload["reason"] = "破壊的なコマンドは PreToolUse hook でブロックされました"
    return payload


def run_agent(task: str, sandbox: str) -> dict[str, Any]:
    plan = PlanState(goal=task, items=[])
    budget = Budget()
    hooks = HookBus()
    trace: list[dict[str, Any]] = []

    hooks.on("PreToolUse", destructive_guard)
    hooks.on("PostToolUse", lambda p: (trace.append({"event": "tool", **p}), p)[1])
    hooks.on("SessionStart", lambda p: (trace.append({"event": "start", **p}), p)[1])
    hooks.on("SessionEnd", lambda p: (trace.append({"event": "end", **p}), p)[1])

    hooks.fire("SessionStart", {"task": task, "sandbox": sandbox,
                                "started_at": time.time()})

    turn = 0
    while True:
        stop = budget.exceeded()
        if stop:
            hooks.fire("Stop", {"reason": stop, "turn": turn})
            break

        step = model_step(plan, turn)
        plan.items = step["plan"]
        budget.step(step["tokens"], step["cost"])

        call = step["tool"]
        if call is None:
            hooks.fire("Stop", {"reason": "complete", "turn": turn})
            break

        name, args = call
        pre = hooks.fire("PreToolUse", {"tool": name, "args": args})
        if pre.get("blocked"):
            hooks.fire("PostToolUse", {"tool": name, "blocked": True,
                                       "reason": pre.get("reason", "")})
            turn += 1
            continue

        try:
            result = TOOLS[name](sandbox, **args)
            hooks.fire("PostToolUse", {"tool": name, "ok": True,
                                       "bytes": len(result)})
        except Exception as exc:
            hooks.fire("PostToolUse", {"tool": name, "ok": False,
                                       "error": str(exc)})

        turn += 1

    hooks.fire("SessionEnd", {"turns": budget.turns_used,
                              "tokens": budget.tokens_used,
                              "dollars": budget.dollars_used})

    return {"plan": plan.summary(), "budget": asdict(budget), "trace": trace}


def main() -> None:
    task = "network call なしで plan-act-observe loop をデモする"
    sandbox = os.path.dirname(os.path.abspath(__file__))
    result = run_agent(task, sandbox)
    print(result["plan"])
    print("---")
    print(f"turns={result['budget']['turns_used']} "
          f"tokens={result['budget']['tokens_used']} "
          f"dollars=${result['budget']['dollars_used']:.3f}")
    print("---")
    print(f"trace event 数: {len(result['trace'])}")
    for ev in result["trace"]:
        print(" ", json.dumps(ev, default=str))


if __name__ == "__main__":
    main()
